src/node_serdes.cc

```diff
@@ -0,0 +1,457 @@
+#include "node.h"
+#include "node_buffer.h"
+#include "base-object.h"
+#include "base-object-inl.h"
+#include "env.h"
+#include "env-inl.h"
+#include "v8.h"
+
+namespace node {
+
+using v8::Array;
+using v8::ArrayBuffer;
+using v8::Context;
+using v8::Function;
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::Integer;
+using v8::Isolate;
+using v8::Just;
+using v8::Local;
+using v8::Maybe;
+using v8::MaybeLocal;
+using v8::Nothing;
+using v8::Object;
+using v8::SharedArrayBuffer;
+using v8::String;
+using v8::Value;
+using v8::ValueDeserializer;
+using v8::ValueSerializer;
+
+class SerializerContext : public BaseObject,
+                          public ValueSerializer::Delegate {
+ public:
+  SerializerContext(Environment* env,
+                    Local<Object> wrap,
+                    Local<Value> throwDataCloneError,
+                    Local<Value> writeHostObject);
+
+  ~SerializerContext() {}
+
+  void ThrowDataCloneError(Local<String> message) override;
+  Maybe<bool> WriteHostObject(Isolate* isolate, Local<Object> object) override;
+
+  static void New(const FunctionCallbackInfo<Value>& args);
+  static void WriteHeader(const FunctionCallbackInfo<Value>& args);
+  static void WriteValue(const FunctionCallbackInfo<Value>& args);
+  static void ReleaseBuffer(const FunctionCallbackInfo<Value>& args);
+  static void TransferArrayBuffer(const FunctionCallbackInfo<Value>& args);
+  static void WriteUint32(const FunctionCallbackInfo<Value>& args);
+  static void WriteUint64(const FunctionCallbackInfo<Value>& args);
+  static void WriteDouble(const FunctionCallbackInfo<Value>& args);
+  static void WriteRawBytes(const FunctionCallbackInfo<Value>& args);
+ private:
+  ValueSerializer serializer_;
+};
+
+class DeserializerContext : public BaseObject,
+                            public ValueDeserializer::Delegate {
+ public:
+  DeserializerContext(Environment* env,
+                      Local<Object> wrap,
+                      Local<Value> buffer,
+                      Local<Value> readHostObject);
+
+  ~DeserializerContext() {}
+
+  MaybeLocal<Object> ReadHostObject(Isolate* isolate) override;
+
+  static void New(const FunctionCallbackInfo<Value>& args);
+  static void ReadHeader(const FunctionCallbackInfo<Value>& args);
+  static void ReadValue(const FunctionCallbackInfo<Value>& args);
+  static void TransferArrayBuffer(const FunctionCallbackInfo<Value>& args);
+  static void GetWireFormatVersion(const FunctionCallbackInfo<Value>& args);
+  static void ReadUint32(const FunctionCallbackInfo<Value>& args);
+  static void ReadUint64(const FunctionCallbackInfo<Value>& args);
+  static void ReadDouble(const FunctionCallbackInfo<Value>& args);
+  static void ReadRawBytes(const FunctionCallbackInfo<Value>& args);
```

Comment:
> Consider giving all methods `inline` linkage if they're not used outside this file.

These functions are passed to V8, so what at least gcc does is to expose them as weak symbols instead of regular symbols.

I can wrap everything in here in an anonymous namespace if you want, if it’s about hiding them – is there a reason we don’t do that in our other source files?